cmd/docker-proxy/network_proxy_test.go

```diff
@@ -167,20 +182,17 @@ func testProxyAt(t *testing.T, proto string, proxy Proxy, addr string, halfClose
 	}
 }
 
-func testProxy(t *testing.T, proto string, proxy Proxy, halfClose bool) {
-	testProxyAt(t, proto, proxy, proxy.FrontendAddr().String(), halfClose)
-}
-
 func testTCP4Proxy(t *testing.T, halfClose bool) {
 	backend := NewEchoServer(t, "tcp", "127.0.0.1:0", EchoServerOptions{TCPHalfClose: halfClose})
 	defer backend.Close()
 	backend.Run()
 	frontendAddr := &net.TCPAddr{IP: net.IPv4(127, 0, 0, 1), Port: 0}
-	proxy, err := NewProxy(frontendAddr, backend.LocalAddr())
+	listenFd, frontendAddr := tcpListener(t, "tcp4", frontendAddr)
```

Comment:
> `listenFd` isn't an 'raw' fd, it's a `*net.TCPListener`. And this makes me think that there's no tests making sure 'fd passing' (and reconstructing a listener out of it) works properly. Maybe we should have a dedicated tests 🤔

It's called `listener` now.

> On the other hand, we have `TestTCPDualStackProxy` which isn't realistic IMO. Maybe we should put everything after `parseFlags` in proxy's `main.go` into a separate func, and write tests in such a way that the proxy is started with a valid config. However, this might make this PR even more heavyweight, so maybe it's fine to do that in a follow-up.

Yes, it's worth using `newProxy` to create the `Proxy` from a `ProxyConfig`. Then, as well as checking the Go listener can be reconstructed from the socket, it can test `newProxy` creating its own TCP/UDP listener (for the case where one isn't passed in by an old version of the daemon). That's not a big change to the tests, can do it here.

Splitting `main` after the call to `parseFlags` would be more involved - it'd mean dealing with the expected pipe on `fd:3`, it starts the proxy as well as creating it, and it starts a goroutine to deal with `SIGTERM`. That stuff hasn't changed, I don't think it's worth the extra upheaval here.